src/ucm/bistro/bistro_loongarch64.c

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (C) Xing Li, Dandan Zhang, 2024. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#if defined(__loongarch64)
+
+#include <ucs/arch/cpu.h>
+#include <ucm/bistro/bistro.h>
+#include <ucm/bistro/bistro_int.h>
+#include <ucs/debug/assert.h>
+#include <ucs/sys/math.h>
+#include <ucm/util/sys.h>
+
+#include <assert.h>
+#include <dlfcn.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+
+#define T0 12
+#define T2 14
+#define RA  1
+#define ZERO  0
+
+/**
+  * @brief JIRL - Add 16 bit immediate to source register, save to destination
+  * register, jump and link from destination register
+  *
+  * @param[in] _regd source register number (0-31)
+  * @param[in] _regj destination register number (0-31)
+  * @param[in] _imm 16 bit immmediate value
+  */
+#define JIRL(_regd, _regj, _imm) \
+    (((0x13) << 26 ) | ((_imm) << 10) | ((_regj) << 5) | (_regd))
+/**
+  * @brief B - Indirect jump
+  *
+  * @param[in] _imm 26 bit immmediate value
+  */
+#define B(_imm) \
+	((0x14) << 26) | (((_imm) & 0xffff) << 10) | ((_imm) >>16)
+
+/**
+  * @brief PCADDU12I - Add upper intermediate to PC
+  *
+  * @param[in] _regd register number (0-31)
+  * @param[in] _imm 20 bit immmediate value
+  */
+#define PCADDU12I(_regd, _imm) (((0xe) << 25) | ((_imm) << 5) | (_regd))
```

Comment:
I'd suggest using `pcalau12i` for forming the upper bits of PC-relative addresses, because `pcalau12i + ld` is able to express some offset ranges that `pcaddu12i + ld` cannot, for example `PC + [0x7ffff800, 0x7fffffff]` (`ld` offsets are treated as signed, so to compensate that, the `pcaddu12i` immediate needs to go beyond `0x7ffff` but that would be a signed overflow).